pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/mergerollup/MergeRollupTaskGenerator.java

```diff
@@ -61,13 +63,26 @@
 /**
  * A {@link PinotTaskGenerator} implementation for generating tasks of type {@link MergeRollupTask}
  *
- * TODO: Add the support for realtime table
+ * Assumptions:
+ *  - When the MergeRollupTask starts the first time, records older than the bufferTimeMs have already been ingested.
```

Comment:
Currently, we make sure `bucketEndMs <= now - bufferTimeMs` to process the time bucket. What if we use maxEndTimeMs of valid segments instead of now? Will it help the bootstrap and pause/resume case?

Reply:
`getValidBucketEndTimeMsForSegment` is only used for the delay metrics. When we validate the target bucket in `isValidBucketEndTime`, we do the following:
```
    // Check that bucketEndMs <= now - bufferMs
    if (bucketEndMs > System.currentTimeMillis() - bufferMs) {
      return false;
    }
```
If we use maxEndTimeMs of segments instead of `now`, we won't have the issue that `latest watermarks advanced too far before records are ingested`?

Reply:
1. I agree that the code I pointed out is only used by metrics
2. the code snippet you posted here is used to prevent bucketEndMs from moving too far

The real progress is managed by the `watermark`, it will not move too far: https://github.com/apache/pinot/blob/master/pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/mergerollup/MergeRollupTaskGenerator.java#L342. This value will not excceed the max end time ms of all ready to process segments.

What do you think?

Reply:
Good point, got it now after rethinking about this. Yes, the assumption lgtm now.